Provides full unicode output for wincon

okibcn opened the pull request:
Windows API requires a sequence of UTF-16 encoded wchars. This commit identifies whether a Unicode codepoint is above 0xFFFF and, in that case, encodes them into UTF-16 before passing it to windows API.

Bill-Gray commented:
Thank you;  this looks quite good.  I expect to pull it shortly,  possibly just 'as is',  though we should look into a few things : 

- At line 640 in `getch.c`,  I don't see this as just a Win32 issue.  I _think_ we can just remove that,  but will need to test.  (I've been _very_ short on time lately,  which is why pull requests and issues have been piling up...)
-  `wget_wch()` should now call `raw_wgetch()`,  not `wgetch()`.
- You're definitely right that `SP->key_modifiers` shouldn't be zeroed every time we call `PDC_get_key()`.  However,  the modifiers apply to key _and_ mouse events,  and I think have to be zeroed both for `KEY_EVENT` (the bit you did) _and_ for `MOUSE_EVENT`.  Again,  I'll have to test.
- In display of SMP characters in WinCon,  the expansion of one character into a Unicode surrogate pair will eat up the buffer twice as fast.  Easy enough to fix;  we'll just have to make `buffer` be twice as big in wide-character mode.  The logic you've applied there can also be used in `_show_run_of_nonansi_characters()`.

okibcn commented:
No problem, take your time. Regarding the points you mention:

> At line 640 in getch.c, I don't see this as just a Win32 issue. I think we can just remove that, but will need to test.

Well it all depends on the intended result in char and wchar mode. That discussion permeates all the code. Probably we should open a discussion about it in the discussions section.

> wget_wch() should now call raw_wgetch(), not wgetch().

In my understanding, get_wch() should return the full 20 bits unicode. At least that is the intended functionality in Linux. But for that we need 32 bits wchar_t in windows like Linux does. And we would need to decouple the wchar_t used in tie common and pd curses folders, from the environment layer that should interface with the Windows API with 16 bits WCHAR.

> You're definitely right that SP->key_modifiers shouldn't be zeroed every time we call PDC_get_key(). However, the modifiers apply to key and mouse events, and I think have to be zeroed both for KEY_EVENT (the bit you did) and for MOUSE_EVENT. Again, I'll have to test.

I also thought about it. Resetting the modifiers actually should be done just before they are required to be updated, probably inside _process_key_event() and inside _process_mouse_event() if there is a button mouse event. But the change now mimics the behavior in the wingui environment.

> In display of SMP characters in WinCon, the expansion of one character into a Unicode surrogate pair will eat up the buffer twice as fast. Easy enough to fix; we'll just have to make buffer be twice as big in wide-character mode. The logic you've applied there can also be used in _show_run_of_nonansi_characters().

Well, non ANSI characters can be multi-word? I thought they were function codes that will never be going to be DWORD. This is why I didn't modify that function. If I am wrong, then of course we could implement the same patch there.



Other than that, I had to force the use of 32 bits wchar_t in windows for it to work properly in my nano For Windows project. ncurses did something similar but they took the long route to adapt it to the 16 bits wchar_t in Windows API. I would like to discuss the possibility of using always 21 bits wchar_t, always storing unicodes. And having the WCHAR type as the 16 bits for API interfaces with Windows. I believe there is a huge mess trying to use the UTF-16 wchar_t concept of the Windows API as the 32 bits wchar_t originally envisioned in the Linux environment. In linux it always stores a unicode. For instance the 😀 will always be a a chtype with the attributes and colors, a wchar_t containing 0x1F600, a sequence of 2 WCHAR with the UTF-16 encoding, or a sequence of 4 chars wit the UTF-8 encoding.

I got into a lot of trouble with mbtowc and other similar functions when using 16 bits wchar_t. I believe that global disambiguation of types could clarify a lot, and simplify the code.

okibcn commented:
In the last commit, I added full Unicode support for non-ansi consoles such as Windows Terminal.

I also corrected when to fall back to ansi mode. WT supports full mode with transparent background that was not enabled properly.

okibcn commented:
Also duplicated the buffer as suggested in your [comment](https://github.com/Bill-Gray/PDCursesMod/pull/273#issuecomment-1407849885).

GitMensch commented:
Can you please update the changelog (docs/HISTORY.md, too?
On the one side it help during reviewing to see the key points and then it also helps the users to know what was done / is happening - without the maintainer to do that work.

GitMensch commented on wincon/Makefile:
```diff
@@ -55,10 +55,12 @@ endif
 # Only decision is:  are we doing a 64-bit compile (_w64 defined)?
 
 ifndef ON_WINDOWS
+    PREFIX  = i686-w64-mingw32-
```
please use consistent spacing, so in this case a tab

okibcn replied:
I copy-n-pasted the other lines, I fixed the modified lines.

GitMensch commented:
The last two changes are not related to the original topic. Can you please send a separate PR with those (and inspecting if wingui and the other ports may need additional changes) and drop them ("git reset --soft HEAD~" (for the docs) then "git reset --har HEAD~2" to drop the other two, then force-push)

okibcn commented:
@GitMensch I have updated the [history.md](https://github.dev/okibcn/PDCursesMod.fork/blob/a6eb477c0f3d719b6b85b090afd963c60c7210ed/docs/HISTORY.md#L25) according to the improvements in this PR.

okibcn commented:
> The last two changes are not related to the original topic. Can you please send a separate PR with those (and inspecting if wingui and the other ports may need additional changes) and drop them ("git reset --soft HEAD~" (for the docs) then "git reset --har HEAD~2" to drop the other two, then force-push)

Oops !!!! I think I screwed it up. I am not good with git, I will revert the Makefile changes with another commit and will create another PR for the makefile changes. I don't have a local copy of this, everything is done in GitHub.

okibcn commented:
> The last two changes are not related to the original topic. Can you please send a separate PR with those (and inspecting if wingui and the other ports may need additional changes) and drop them ("git reset --soft HEAD~" (for the docs) then "git reset --har HEAD~2" to drop the other two, then force-push)

@GitMensch I have removed the Makefile modifications and the related annotations in History. I am creating a separate PR for that.

GitMensch commented:
> I don't have a local copy of this, everything is done in GitHub.

You may want to have a look at https://gitpod.io#github.com/Bill-Gray/PDCursesMod/pull/273 - I really enjoy working with it.

okibcn commented:
> > I don't have a local copy of this, everything is done in GitHub.
> 
> You may want to have a look at https://gitpod.io#github.com/Bill-Gray/PDCursesMod/pull/273 - I really enjoy working with it.

Thanks! I will explore this. It resembles GitHub codespaces.

GitMensch commented:
> It resembles GitHub codespaces.

Actually it predates those by years ;-)

Bill-Gray commented:
Just got back to this.  Things did get a little confused,  with some pieces of this pull request appearing as commits in other pull requests;  I had to do some cherry-picking.

 So I checked each commit against the current commit log.  All of the changes in this pull request _have_ been pulled,  though it was done in bits and pieces.  So I think we can close this one out.  Thank you;  it's some nice improvements to the code... should be appreciated by the WinCon and Windows on Arm folks out there.

GitMensch commented:
Thank for re-checking and ❤️ to @okibcn for the original work.

@Bill-Gray: please update HISTORY.md with giving credit.

> should be appreciated by the WinCon and Windows on Arm folks out there.

Very likely. Is that enough for a new release?

Bill-Gray commented:
> Very likely. Is that enough for a new release?

I'm open to doing a release at basically any time.  It's not difficult to do,  so the usual reason for a release is "because somebody asked me to".

Other than the usual updates to `HISTORY.md` and the version constants in `curses.h`,  is there anything anyone feels Must Be Done before a new release?  I'd consider adding in Julius' OpenGL port (pull #284),  for example.

GitMensch commented:
Can you possibly amend the cherry-picked commit messages to reference the issues they fix? I find that additional information _really_ helpful.
... in the meantime I'm closing and referencing those from the issue.

GitMensch commented:
Is #274 really fixed now? There seems to be no cherry-pick of [the fixing commit](https://github.com/Bill-Gray/PDCursesMod/pull/273/commits/7e28f97cf158c33bf4944beab19d9e09a6ef3277).

GitMensch commented:
Not a "must be done", but what about #283 which had two working prototypes?

Bill-Gray commented:
Re issue #274 : that one got fixed with commit 20507066497.  I'm afraid things got slightly muddled here... that was partly why I had to do some cherry-picking of commits for things that hadn't been fixed elsewhere and which wouldn't conflict with something else.

I _think_ (but should double-check) that we have fixed the various issues underlying #283.  (Specifically,  `getch()` is basically an eight-bit solution -- hit,  say,  £ or ä and you should get three bytes corresponding to the UTF-8 rendering of that Unicode point;  `wget_wch()` can just put the Unicode point into a `wint_t`.)

There's one hitch to all this,  though : what do you do if `wint_t` is only 16 bits?  You're unable to return an SMP value in that case.

In the Microsoft and OpenWATCOM compilers,  `sizeof(wint_t) == 2`.   On Microsoft platforms,  it appears `wget_wch()` will have to return two Unicode surrogates and the program will have to combine them... or we'll have to provide a `wget_unicode_point()` function,  which would say something like
```c
int wget_unicode_point( WINDOW *win, int32_t *cp)
{
   wint_t wch, wch2;
  int rval = get_wch( win, &wch);

   if( ERR != rval)
      {
      if( IS_HIGH_SURROGATE( wch))
         {
         rval = get_wch( win, &wch2);
         if( ERR != rval && IS_LOW_SURROGATE( wch2))
            *cp = ((int32_t)(wch - 0xd800) << 10) + 0x10000 + (int32_t)wch2 - 0xdc00;
         }
      else
         *cp = (int32_t)wch;
      }
   return( rval);
}
```
I would suggest that this function should _not_ be part of PDCursesMod,  because it would be an issue for _any_ curses implementation under MSWin.  You'd put the function in your program and then be able to link to PDCurses or ncurses or PDCursesMod and get SMP key entries.

GitMensch commented:
... so is #283 fixed now (the it should be closed and possibly referenced in HISTORY), is the comment above something that should be added to documentation somewhere to be able to close that?
